File: apps/web/lib/api/business.ts
import { apiFetch } from "./client";
import { getToken } from "@/lib/auth/session";

function authHeaders(): HeadersInit {
  const token = getToken();
  return token ? { Authorization: `Bearer ${token}` } : {};
}

export type Stage = "idea" | "validation" | "planning" | "pre_launch" | "launched" | "operating";
export type DeliveryMode = "online" | "offline" | "hybrid";
export type TaskStatus = "pending" | "in_progress" | "completed";
export type TaskPriority = "low" | "medium" | "high";
export type EntryType = "revenue" | "expense";
export type Confidence = "Low" | "Medium" | "High";

export interface BusinessProject {
  id: number;
  name: string;
  description: string | null;
  industry: string | null;
  location: string | null;
  stage: Stage;
  startup_budget: string | null;
  currency: string;
  target_customer: string | null;
  business_model: string | null;
  experience_level: string | null;
  delivery_mode: DeliveryMode | null;
  time_commitment: string | null;
  goals: string | null;
  created_at: string;
  updated_at: string;
}

export interface BusinessProjectInput {
  name: string;
  description?: string;
  industry?: string;
  location?: string;
  startup_budget?: string;
  target_customer?: string;
  business_model?: string;
  experience_level?: string;
  delivery_mode?: DeliveryMode;
  time_commitment?: string;
  goals?: string;
}

export function createBusiness(input: BusinessProjectInput): Promise<BusinessProject> {
  return apiFetch<BusinessProject>("/businesses", {
    method: "POST",
    headers: authHeaders(),
    body: JSON.stringify(input),
  });
}

export function listBusinesses(): Promise<BusinessProject[]> {
  return apiFetch<BusinessProject[]>("/businesses", { headers: authHeaders() });
}

export function getBusiness(id: number): Promise<BusinessProject> {
  return apiFetch<BusinessProject>(`/businesses/${id}`, { headers: authHeaders() });
}

export function updateBusiness(id: number, patch: Partial<BusinessProjectInput & { stage: Stage }>): Promise<BusinessProject> {
  return apiFetch<BusinessProject>(`/businesses/${id}`, {
    method: "PATCH",
    headers: authHeaders(),
    body: JSON.stringify(patch),
  });
}

export interface ValidationScores {
  demand: number;
  differentiation: number;
  business_model: number;
  capital: number;
  operational: number;
  gtm: number;
  total: number;
}

export interface ValidationReportSections {
  business_summary: string;
  target_customer: string;
  problem_being_solved: string;
  revenue_model: string;
  startup_requirements: string;
  estimated_cost_areas: string[];
  operational_complexity: string;
  competition_considerations: string;
  differentiation_opportunities: string;
  major_risks: string[];
  questions_to_validate: string[];
  recommended_next_actions: string[];
}

export interface ValidationReport {
  id: number;
  business_project_id: number;
  scores: ValidationScores;
  confidence: Confidence;
  verdict: string;
  missing_info: string[];
  sections: ValidationReportSections;
  created_at: string;
}

export function validateBusiness(id: number): Promise<ValidationReport> {
  return apiFetch<ValidationReport>(`/businesses/${id}/validate`, { method: "POST", headers: authHeaders() });
}

export function getLatestValidation(id: number): Promise<ValidationReport> {
  return apiFetch<ValidationReport>(`/businesses/${id}/validation`, { headers: authHeaders() });
}

export interface BudgetItem {
  id: number;
  category: string;
  amount: string;
  notes: string | null;
}

export interface BudgetItemInput {
  category: string;
  amount: string;
  notes?: string;
}

export interface Budget {
  items: BudgetItem[];
  total: string;
}

export function getBudget(businessId: number): Promise<Budget> {
  return apiFetch<Budget>(`/businesses/${businessId}/budget`, { headers: authHeaders() });
}

export function putBudget(businessId: number, items: BudgetItemInput[]): Promise<Budget> {
  return apiFetch<Budget>(`/businesses/${businessId}/budget`, {
    method: "PUT",
    headers: authHeaders(),
    body: JSON.stringify({ items }),
  });
}

export interface BusinessTask {
  id: number;
  title: string;
  description: string | null;
  category: string | null;
  status: TaskStatus;
  priority: TaskPriority;
  week: number | null;
  order_index: number;
  due_date: string | null;
  created_at: string;
  updated_at: string;
}

export function listTasks(businessId: number): Promise<BusinessTask[]> {
  return apiFetch<BusinessTask[]>(`/businesses/${businessId}/tasks`, { headers: authHeaders() });
}

export function createTask(
  businessId: number,
  input: { title: string; description?: string; category?: string; priority?: TaskPriority; week?: number }
): Promise<BusinessTask> {
  return apiFetch<BusinessTask>(`/businesses/${businessId}/tasks`, {
    method: "POST",
    headers: authHeaders(),
    body: JSON.stringify(input),
  });
}

export function updateTask(businessId: number, taskId: number, patch: { status?: TaskStatus }): Promise<BusinessTask> {
  return apiFetch<BusinessTask>(`/businesses/${businessId}/tasks/${taskId}`, {
    method: "PATCH",
    headers: authHeaders(),
    body: JSON.stringify(patch),
  });
}

export function generateRoadmap(businessId: number): Promise<BusinessTask[]> {
  return apiFetch<BusinessTask[]>(`/businesses/${businessId}/tasks/generate-roadmap`, {
    method: "POST",
    headers: authHeaders(),
  });
}

export interface FinancialEntry {
  id: number;
  type: EntryType;
  amount: string;
  category: string | null;
  description: string | null;
  entry_date: string;
  created_at: string;
}

export function listFinancials(businessId: number): Promise<FinancialEntry[]> {
  return apiFetch<FinancialEntry[]>(`/businesses/${businessId}/financials`, { headers: authHeaders() });
}

export function createFinancialEntry(
  businessId: number,
  input: { type: EntryType; amount: string; category?: string; description?: string; entry_date: string }
): Promise<FinancialEntry> {
  return apiFetch<FinancialEntry>(`/businesses/${businessId}/financials`, {
    method: "POST",
    headers: authHeaders(),
    body: JSON.stringify(input),
  });
}

export interface FinancialSummary {
  total_revenue: string;
  total_expenses: string;
  net_result: string;
  expense_breakdown: { category: string; total: string }[];
  entry_count: number;
}

export function getFinancialSummary(businessId: number): Promise<FinancialSummary> {
  return apiFetch<FinancialSummary>(`/businesses/${businessId}/financials/summary`, { headers: authHeaders() });
}

// ---- Calculators (stateless, no auth) --------------------------------------

export interface BreakEvenResult {
  contribution_per_unit: string;
  break_even_units: string | null;
  break_even_revenue: string | null;
  is_viable: boolean;
  explanation: string;
}

export function calculateBreakEven(input: {
  fixed_costs: string;
  selling_price: string;
  variable_cost_per_unit: string;
}): Promise<BreakEvenResult> {
  return apiFetch<BreakEvenResult>("/calculators/break-even", { method: "POST", body: JSON.stringify(input) });
}

export interface MarginResult {
  gross_profit: string;
  gross_margin_pct: string | null;
  operating_profit: string;
  operating_margin_pct: string | null;
  explanation: string;
}

export function calculateMargin(input: {
  revenue: string;
  cogs: string;
  operating_expenses: string;
}): Promise<MarginResult> {
  return apiFetch<MarginResult>("/calculators/margin", { method: "POST", body: JSON.stringify(input) });
}

export interface PricingResult {
  suggested_price: string;
  profit_per_unit: string;
  explanation: string;
}

export function calculatePricing(input: {
  cost_per_unit: string;
  desired_margin_pct: string;
  fees_pct?: string;
}): Promise<PricingResult> {
  return apiFetch<PricingResult>("/calculators/pricing", { method: "POST", body: JSON.stringify(input) });
}

export interface WhatIfResult {
  current_profit: string;
  projected_profit: string;
  change: string;
  explanation: string;
}

export function calculateWhatIf(input: {
  current_monthly_profit: string;
  additional_monthly_cost?: string;
  additional_monthly_revenue?: string;
}): Promise<WhatIfResult> {
  return apiFetch<WhatIfResult>("/calculators/what-if", { method: "POST", body: JSON.stringify(input) });
}

File: apps/api/app/schemas/business.py
from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

Stage = Literal["idea", "validation", "planning", "pre_launch", "launched", "operating"]
TaskStatus = Literal["pending", "in_progress", "completed"]
TaskPriority = Literal["low", "medium", "high"]
EntryType = Literal["revenue", "expense"]
DeliveryMode = Literal["online", "offline", "hybrid"]
Confidence = Literal["Low", "Medium", "High"]


# ---- Business Project -----------------------------------------------------


class BusinessProjectCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=4000)
    industry: str | None = Field(default=None, max_length=120)
    location: str | None = Field(default=None, max_length=120)
    startup_budget: Decimal | None = Field(default=None, ge=0)
    currency: str = Field(default="INR", min_length=3, max_length=3)
    target_customer: str | None = Field(default=None, max_length=2000)
    business_model: str | None = Field(default=None, max_length=2000)
    experience_level: str | None = Field(default=None, max_length=40)
    delivery_mode: DeliveryMode | None = None
    time_commitment: str | None = Field(default=None, max_length=40)
    goals: str | None = Field(default=None, max_length=2000)


class BusinessProjectUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=4000)
    industry: str | None = Field(default=None, max_length=120)
    location: str | None = Field(default=None, max_length=120)
    stage: Stage | None = None
    startup_budget: Decimal | None = Field(default=None, ge=0)
    target_customer: str | None = Field(default=None, max_length=2000)
    business_model: str | None = Field(default=None, max_length=2000)
    experience_level: str | None = Field(default=None, max_length=40)
    delivery_mode: DeliveryMode | None = None
    time_commitment: str | None = Field(default=None, max_length=40)
    goals: str | None = Field(default=None, max_length=2000)


class BusinessProjectRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    description: str | None
    industry: str | None
    location: str | None
    stage: Stage
    startup_budget: Decimal | None
    currency: str
    target_customer: str | None
    business_model: str | None
    experience_level: str | None
    delivery_mode: str | None
    time_commitment: str | None
    goals: str | None
    created_at: datetime
    updated_at: datetime


# ---- Idea Validation --------------------------------------------------------


class ValidationScores(BaseModel):
    demand: int = Field(ge=0, le=20)
    differentiation: int = Field(ge=0, le=15)
    business_model: int = Field(ge=0, le=15)
    capital: int = Field(ge=0, le=20)
    operational: int = Field(ge=0, le=15)
    gtm: int = Field(ge=0, le=15)
    total: int = Field(ge=0, le=100)


class ValidationReportSections(BaseModel):
    """The narrative parts of a validation report. Populated by the
    deterministic engine from structured inputs, then optionally enriched
    by the LLM's structured analysis (see app/services/validation_engine.py)
    — never freeform, never trusted un-validated from the model."""

    business_summary: str
    target_customer: str
    problem_being_solved: str
    revenue_model: str
    startup_requirements: str
    estimated_cost_areas: list[str] = Field(default_factory=list)
    operational_complexity: str
    competition_considerations: str
    differentiation_opportunities: str
    major_risks: list[str] = Field(default_factory=list)
    questions_to_validate: list[str] = Field(default_factory=list)
    recommended_next_actions: list[str] = Field(default_factory=list)


class ValidationReport(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    business_project_id: int
    scores: ValidationScores
    confidence: Confidence
    verdict: str
    missing_info: list[str]
    sections: ValidationReportSections
    created_at: datetime

    @classmethod
    def from_model(cls, model) -> "ValidationReport":  # type: ignore[no-untyped-def]
        report = model.report
        return cls(
            id=model.id,
            business_project_id=model.business_project_id,
            scores=ValidationScores(
                demand=model.score_demand,
                differentiation=model.score_differentiation,
                business_model=model.score_business_model,
                capital=model.score_capital,
                operational=model.score_operational,
                gtm=model.score_gtm,
                total=model.total_score,
            ),
            confidence=model.confidence,
            verdict=model.verdict,
            missing_info=report.get("missing_info", []),
            sections=ValidationReportSections(**report["sections"]),
            created_at=model.created_at,
        )


# ---- Budget -----------------------------------------------------------------


class BudgetItemCreate(BaseModel):
    category: str = Field(min_length=1, max_length=120)
    amount: Decimal = Field(ge=0)
    notes: str | None = Field(default=None, max_length=500)


class BudgetItemRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    category: str
    amount: Decimal
    notes: str | None


class BudgetPut(BaseModel):
    """Replaces the entire budget item list — the frontend calculator
    always edits the full set, so a full replace avoids partial-update
    drift between categories."""

    items: list[BudgetItemCreate] = Field(max_length=50)


class BudgetRead(BaseModel):
    items: list[BudgetItemRead]
    total: Decimal


# ---- Tasks --------------------------------------------------------------


class BusinessTaskCreate(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=2000)
    category: str | None = Field(default=None, max_length=60)
    priority: TaskPriority = "medium"
    week: int | None = Field(default=None, ge=1, le=52)
    due_date: date | None = None


class BusinessTaskUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=2000)
    status: TaskStatus | None = None
    priority: TaskPriority | None = None
    due_date: date | None = None


class BusinessTaskRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    description: str | None
    category: str | None
    status: TaskStatus
    priority: TaskPriority
    week: int | None
    order_index: int
    due_date: date | None
    created_at: datetime
    updated_at: datetime


# ---- Financials -----------------------------------------------------------


class FinancialEntryCreate(BaseModel):
    type: EntryType
    amount: Decimal = Field(gt=0)
    category: str | None = Field(default=None, max_length=120)
    description: str | None = Field(default=None, max_length=500)
    entry_date: date


class FinancialEntryRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    type: EntryType
    amount: Decimal
    category: str | None
    description: str | None
    entry_date: date
    created_at: datetime


class ExpenseCategoryTotal(BaseModel):
    category: str
    total: Decimal


class FinancialSummary(BaseModel):
    total_revenue: Decimal
    total_expenses: Decimal
    net_result: Decimal
    expense_breakdown: list[ExpenseCategoryTotal]
    entry_count: int


# ---- Calculators ------------------------------------------------------------


class BreakEvenRequest(BaseModel):
    fixed_costs: Decimal = Field(ge=0)
    selling_price: Decimal = Field(ge=0)
    variable_cost_per_unit: Decimal = Field(ge=0)


class BreakEvenResponse(BaseModel):
    contribution_per_unit: Decimal
    break_even_units: Decimal | None
    break_even_revenue: Decimal | None
    is_viable: bool
    explanation: str


class MarginRequest(BaseModel):
    revenue: Decimal = Field(ge=0)
    cogs: Decimal = Field(ge=0)
    operating_expenses: Decimal = Field(ge=0)


class MarginResponse(BaseModel):
    gross_profit: Decimal
    gross_margin_pct: Decimal | None
    operating_profit: Decimal
    operating_margin_pct: Decimal | None
    explanation: str


class PricingRequest(BaseModel):
    cost_per_unit: Decimal = Field(ge=0)
    desired_margin_pct: Decimal = Field(ge=0, lt=100)
    fees_pct: Decimal = Field(default=Decimal("0"), ge=0, lt=100)


class PricingResponse(BaseModel):
    suggested_price: Decimal
    profit_per_unit: Decimal
    explanation: str


class WhatIfRequest(BaseModel):
    current_monthly_profit: Decimal
    additional_monthly_cost: Decimal = Field(default=Decimal("0"), ge=0)
    additional_monthly_revenue: Decimal = Field(default=Decimal("0"), ge=0)


class WhatIfResponse(BaseModel):
    current_profit: Decimal
    projected_profit: Decimal
    change: Decimal
    explanation: str
